contracts/Governance.sol

```diff
@@ -0,0 +1,925 @@
+    function proposalEta(uint256 proposalId)
+        public
+        view
+        virtual
+        returns (uint256)
+    {
+        uint256 eta = _timelock.getTimestamp(_timelockIds[proposalId]);
+        return eta == 1 ? 0 : eta; // _DONE_TIMESTAMP (1) should be replaced with a 0 value
+    }
+
+    function hashProposal(
+        address[] memory targets,
+        uint256[] memory values,
+        bytes[] memory calldatas,
+        bytes32 descriptionHash
+    ) public pure virtual returns (uint256) {
+        return
+            uint256(
+                keccak256(
+                    abi.encode(targets, values, calldatas, descriptionHash)
+                )
+            );
+    }
+
+    function propose(
+        address[] memory targets,
+        uint256[] memory values,
+        bytes32[] memory roles,
+        string[] memory signatures,
+        bytes[] memory calldatas,
+        bytes32[] memory actions,
+        string memory description
+    ) public virtual returns (uint256) {
+        for (uint256 i = 0; i < roles.length; ++i) {
+            require(
+                votersRoles[roles[i]][_msgSender()],
+                'Governance::propose: Proposer must have proposal roles'
+            );
+        }
+        require(
+            targets.length == values.length,
+            'Governance::propose: invalid proposal length'
+        );
+        require(
+            targets.length == calldatas.length,
+            'Governance::propose: invalid proposal length'
+        );
+        require(
+            targets.length == actions.length,
+            'Governance::propose: invalid proposal length'
+        );
+        require(targets.length > 0, 'Governance::propose: empty proposal');
+
+        bytes32 descriptionHash = keccak256(bytes(description));
+        uint256 proposalId = hashProposal(
+            targets,
+            values,
+            _encodeCalldata(signatures, calldatas),
+            descriptionHash
+        );
+
+        Proposal storage proposal = _proposals[proposalId];
+        require(
+            proposal.voteStart.isUnset(),
+            'Governance::propose: proposal already exists'
+        );
+        require(
+            proposal.descriptionHash == bytes32(0),
+            'Governance::propose: proposal already exists'
+        );
+
+        proposal.proposer = _msgSender();
+        proposal.targets = targets;
+        proposal.values = values;
+        proposal.roles = roles;
+        proposal.calldatas = calldatas;
+        proposal.signatures = signatures;
+        proposal.actions = actions;
+        proposal.descriptionHash = descriptionHash;
+
+        uint64 snapshot = block.number.toUint64() + votingDelay().toUint64();
+        uint64 deadline = snapshot + votingPeriod().toUint64();
+
+        proposal.voteStart.setDeadline(snapshot);
+        proposal.voteEnd.setDeadline(deadline);
+
+        return proposalId;
+    }
+
+    function registerNewRole(bytes32 role) external onlyGovernance {
```

Comment:
How would we go about removing an existing role ...if a role that existed became redundant and was voted out by a proposal?

Reply:
we don't for now, we just don't want on it